### PR TITLE
Fix bpf_link__destroy(), add bpf_link__disconnect()

### LINKED ISSUE #484: Fix bpf_link__destroy(), add bpf_link__disconnect()

### PR DESCRIPTION
**Pull request by saxena-anurag**
Fixes #484 

This PR fixes the following APIs:

1. bpf_link__destroy()
2. bpf_program__attach()
3. bpf_program__attach_xdp(). 

The PR also adds bpf_link__disconnect() API which allows to close link fd without detaching.

PR also exports ebpf_link_close() API from ebpfapi.dll

### CODE REVIEW

File: libs/api/libbpf_program.cpp [hunk 1]
```diff
@@ -261,13 +256,25 @@ bpf_object__unpin_programs(struct bpf_object* obj, const char* path)
     return 0;
 }
```
Comment (shankarseal) [inline libs/api/libbpf_program.cpp:258]: please add doxygen style @brief - not sure if there is a decision on what to do with all these libbpf APIs that are implemented.

File: tests/end_to_end/end_to_end.cpp [hunk 2]
```diff
@@ -637,6 +638,53 @@ TEST_CASE("implicit_detach", "[end_to_end]")
...
+    bpf_object__close(object);
+
+    program_fd = ebpf_fd_invalid;
+    REQUIRE(ebpf_get_next_program(program_fd, &program_fd) == EBPF_SUCCESS);
+    REQUIRE(program_fd == ebpf_fd_invalid);
+
+    // Close link handle (without detaching). This should delete the link
+    // object. ebpf_object_tracking_terminate() which is called when the test
+    // exits checks if all the objects in EC have been deleted.
+    bpf_link__disconnect(link);
```
Comment (shankarseal) [inline tests/end_to_end/end_to_end.cpp:684]: should there be a case when bpf_detach is called?

File: libs/api/libbpf_program.cpp (hunk 1 above)
Comment (dthaler) [reply libs/api/libbpf_program.cpp:258]: I will be adding doxygen comments into include/libbpf.h and include/bpf.h

File: tests/end_to_end/end_to_end.cpp [hunk 3]
```diff
@@ -654,6 +655,53 @@ TEST_CASE("implicit_detach", "[end_to_end]")
     hook.close_link(link);
 }
 
+// This test uses bpf_link__disconnect() and bpf_link__destroy() to test
+// implicit detach.
+TEST_CASE("implicit_detach_2", "[end_to_end]")
+{
+    // This test case does the following:
+    // 1. Close program handle. An implicit detach should happen and program
```
Comment (dthaler) [inline tests/end_to_end/end_to_end.cpp:663]:
```suggestion
    // 1. Close program handle. An implicit detach should happen and the program
```

File: tests/end_to_end/end_to_end.cpp [hunk 4]
```diff
@@ -654,6 +655,53 @@ TEST_CASE("implicit_detach", "[end_to_end]")
...
+    // detach the program from the hook and unload the program.
+    bpf_object__close(object);
+
+    program_fd = ebpf_fd_invalid;
+    REQUIRE(ebpf_get_next_program(program_fd, &program_fd) == EBPF_SUCCESS);
+    REQUIRE(program_fd == ebpf_fd_invalid);
+
+    // Close link handle (without detaching). This should delete the link
+    // object. ebpf_object_tracking_terminate() which is called when the test
+    // exits checks if all the objects in EC have been deleted.
```
Comment (dthaler) [inline tests/end_to_end/end_to_end.cpp:700]:
```suggestion
    // exits checks if all the objects in the execution context have been deleted.
```

File: libs/api/libbpf_program.cpp [hunk 5]
```diff
@@ -261,13 +256,25 @@ bpf_object__unpin_programs(struct bpf_object* obj, const char* path)
...
 
+void
+bpf_link__disconnect(struct bpf_link* link)
```
Comment (dthaler) [inline libs/api/libbpf_program.cpp:260]: move this to libbpf_link.cpp.  (bpf_link__destroy should move there too)

Comment (saxena-anurag) [reply libs/api/libbpf_program.cpp]: Done.

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]

**Pull request merged**